File: app/src/main/java/com/minerdev/greformanager/HouseModifyActivity.java
package com.minerdev.greformanager;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.lifecycle.ViewModelProvider;
import androidx.viewpager.widget.ViewPager;

import com.android.volley.Request;
import com.google.gson.Gson;

public class HouseModifyActivity extends AppCompatActivity {
    private static String mode;
    private final SectionPageAdapter adapter = new SectionPageAdapter(getSupportFragmentManager(),
            FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
    private ImageViewModel imageViewModel;
    private HouseModifyViewModel viewModel;
    private NonSwipeViewPager viewPager;
    private Button button_next;
    private Button button_previous;
    private Button button_save;
    private final ViewPager.OnPageChangeListener onPageChangeListener = new ViewPager.OnPageChangeListener() {
        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

        }

        @Override
        public void onPageSelected(int position) {
            switch (position) {
                case 0:
                    button_previous.setEnabled(false);
                    break;

                case 1:
                case 2:
                    button_previous.setEnabled(true);
                    button_next.setVisibility(View.VISIBLE);
                    button_save.setVisibility(View.GONE);
                    break;

                case 3:
                    button_next.setVisibility(View.GONE);
                    button_save.setVisibility(View.VISIBLE);
                    break;

                default:
                    break;
            }
        }

        @Override
        public void onPageScrollStateChanged(int state) {

        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_house_modify);

        imageViewModel = new ViewModelProvider(this,
                new ViewModelProvider.AndroidViewModelFactory(getApplication())).get(ImageViewModel.class);

        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);

        Intent intent = getIntent();
        mode = intent.getStringExtra("mode");
        House house = null;

        if (mode.equals("add")) {
            actionBar.setTitle("매물 추가");

        } else if (mode.equals("modify")) {
            actionBar.setTitle("매물 정보 수정");
            house = intent.getParcelableExtra("house_value");
        }

        viewModel = new ViewModelProvider(this).get(HouseModifyViewModel.class);
        viewModel.setMode(mode, house);

        button_next = findViewById(R.id.house_modify_button_next);
        button_next.setOnClickListener(v -> toNextPage());

        button_previous = findViewById(R.id.house_modify_button_previous);
        button_previous.setEnabled(false);
        button_previous.setOnClickListener(v -> {
            hideKeyboard();
            viewPager.setCurrentItem(viewPager.getCurrentItem() - 1);
        });

        button_save = findViewById(R.id.house_modify_button_save);
        button_save.setVisibility(View.GONE);
        button_save.setOnClickListener(v -> askSave());

        setViewPager();
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;

            default:
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void finish() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage("저장하지 않고 돌아가시겠습니까?");
        builder.setIcon(R.drawable.ic_round_help_24);
        builder.setPositiveButton("확인", (dialog, which) -> {
            HouseModifyActivity.super.finish();
        });

        builder.setNegativeButton("취소", (dialog, which) -> dialog.dismiss());

        AlertDialog alertDialog = builder.create();
        alertDialog.show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        adapter.getItem(viewPager.getCurrentItem()).onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    public void setViewPager() {
        viewPager = findViewById(R.id.house_modify_viewPager);
        viewPager.addOnPageChangeListener(onPageChangeListener);

        adapter.addFragment(new InfoFragment1(), "매물 정보 입력");
        adapter.addFragment(new InfoFragment2(), "매물 정보 입력");
        adapter.addFragment(new InfoFragment3(), "매물 정보 입력");
        adapter.addFragment(new ImageFragment(), "매물 사진 선택");

        viewPager.setAdapter(adapter);
    }

    private void hideKeyboard() {
        if (getCurrentFocus() != null) {
            InputMethodManager manager = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
            manager.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    private void toNextPage() {
        hideKeyboard();

        int current = viewPager.getCurrentItem();
        OnSaveDataListener listener = (OnSaveDataListener) adapter.getItem(current);
        if (listener.checkData()) {
            listener.saveData();
            viewPager.setCurrentItem(current + 1);

        } else {
            Toast.makeText(getApplicationContext(), "error", Toast.LENGTH_SHORT).show();
        }
    }

    private void askSave() {
        hideKeyboard();

        int current = viewPager.getCurrentItem();
        OnSaveDataListener listener = (OnSaveDataListener) adapter.getItem(current);
        if (listener.checkData()) {
            listener.saveData();

            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setMessage("저장하시겠습니까?");
            builder.setIcon(R.drawable.ic_round_help_24);

            builder.setPositiveButton("확인", (dialog, which) -> {
                if (mode.equals("add")) {
                    add();

                } else if (mode.equals("modify")) {
                    modify();
                }
            });

            builder.setNegativeButton("취소", (dialog, which) -> dialog.dismiss());

            AlertDialog alertDialog = builder.create();
            alertDialog.show();

        } else {
            Toast.makeText(getApplicationContext(), "error", Toast.LENGTH_SHORT).show();
        }
    }

    private void add() {
        HttpConnection.getInstance().send(this, Request.Method.POST,
                "houses", viewModel.getHouse(), receivedData -> {
                    Gson gson = new Gson();
                    House data = gson.fromJson(receivedData, House.class);
                    viewModel.getHouse().id = data.id;
                    viewModel.getHouse().created_at = data.created_at;
                    viewModel.getHouse().updated_at = data.updated_at;

                    HttpConnection.getInstance().send(getApplication(), Request.Method.POST,
                            "houses/" + data.id + "/images", viewModel.getImageUris(), viewModel.getThumbnail(), receivedData1 -> {
                                Gson gson1 = new Gson();
                                Image imageData = gson1.fromJson(receivedData1, Image.class);
                                imageViewModel.insert(imageData);
                                Log.d("DB_INSERT", receivedData1);
                            });

                    Intent intent = new Intent();
                    intent.putExtra("house_value", viewModel.getHouse());
                    setResult(RESULT_OK, intent);
                    HouseModifyActivity.super.finish();
                });
    }

    private void modify() {
        int id = viewModel.getHouse().id;
        HttpConnection.getInstance().send(this, Request.Method.PATCH,
                "houses/" + id, viewModel.getHouse(), receivedData -> {
                    Gson gson = new Gson();
                    House data = gson.fromJson(receivedData, House.class);
                    viewModel.getHouse().updated_at = data.updated_at;

                    HttpConnection.getInstance().send(getApplication(), Request.Method.POST,
                            "houses/" + id + "/images", viewModel.getImageUris(), viewModel.getThumbnail(), receivedData1 -> {
                                Gson gson1 = new Gson();
                                Image imageData = gson1.fromJson(receivedData1, Image.class);
                                imageViewModel.updateOrInsert(imageData);
                                Log.d("DB_INSERT", receivedData1);
                            });

                    Intent intent = new Intent();
                    intent.putExtra("house_value", viewModel.getHouse());
                    setResult(RESULT_OK, intent);
                    HouseModifyActivity.super.finish();
                });
    }
}

File: app/src/main/java/com/minerdev/greformanager/HouseModifyActivity.kt
package com.minerdev.greformanager

import android.content.Context
import android.content.DialogInterface
import android.net.Uri
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.minerdev.greformanager.databinding.ActivityHouseModifyBinding
import java.util.*

class HouseModifyActivity : AppCompatActivity() {
    private val adapter = SectionPageAdapter(supportFragmentManager,
            FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT)
    private val onPageChangeListener: OnPageChangeListener = object : OnPageChangeListener {
        override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

        override fun onPageSelected(position: Int) {
            when (position) {
                0 -> binding.btnPrevious.isEnabled = false
                1, 2 -> {
                    binding.btnPrevious.isEnabled = true
                    binding.btnNext.visibility = View.VISIBLE
                    binding.btnSave.visibility = View.GONE
                }
                3 -> {
                    binding.btnNext.visibility = View.GONE
                    binding.btnSave.visibility = View.VISIBLE
                }
                else -> {
                }
            }
        }

        override fun onPageScrollStateChanged(state: Int) {}
    }
    private val sharedViewModel: SharedViewModel by viewModels()
    private val viewModel: HouseModifyViewModel by viewModels()
    private val binding by lazy { ActivityHouseModifyBinding.inflate(layoutInflater) }

    private var count = 0
    private var mode: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)

        val intent = intent
        mode = intent.getStringExtra("mode")

        if (mode == "add") {
            actionBar?.title = "매물 추가"

        } else if (mode == "modify") {
            actionBar?.title = "매물 정보 수정"

            val house: House = intent.getParcelableExtra("house_value")!!
            sharedViewModel.house = house
            viewModel.getOrderByPosition(sharedViewModel.house.id).observe(this, { images: List<Image> ->
                val uris = ArrayList<Uri>()
                for (image in images) {
                    uris.add(Uri.parse(Constants.instance.DNS + "/storage/images/" + image.house_id + "/" + image.title))
                    if (image.thumbnail.toInt() == 1) {
                        sharedViewModel.thumbnail = image.position.toInt()
                    }
                }

                sharedViewModel.images = images as ArrayList<Image>
                sharedViewModel.imageUris.addAll(uris)
            })
        }

        binding.btnNext.setOnClickListener { toNextPage() }
        binding.btnPrevious.isEnabled = false
        binding.btnPrevious.setOnClickListener {
            hideKeyboard()
            val current = binding.viewPager.currentItem
            val listener = adapter.getItem(current) as OnSaveDataListener
            listener.saveData()
            binding.viewPager.currentItem = binding.viewPager.currentItem - 1
        }
        binding.btnSave.visibility = View.GONE
        binding.btnSave.setOnClickListener { askSave() }
        setViewPager()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun finish() {
        val builder = AlertDialog.Builder(this)
        builder.setMessage("저장하지 않고 돌아가시겠습니까?")
        builder.setIcon(R.drawable.ic_round_help_24)
        builder.setPositiveButton("확인") { _: DialogInterface?, _: Int -> super@HouseModifyActivity.finish() }
        builder.setNegativeButton("취소") { dialog: DialogInterface, _: Int -> dialog.dismiss() }
        val alertDialog = builder.create()
        alertDialog.show()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        adapter.getItem(binding.viewPager.currentItem).onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    private fun setViewPager() {
        binding.viewPager.addOnPageChangeListener(onPageChangeListener)
        adapter.addFragment(InfoFragment1(), "매물 정보 입력")
        adapter.addFragment(InfoFragment2(), "매물 정보 입력")
        adapter.addFragment(InfoFragment3(), "매물 정보 입력")
        adapter.addFragment(ImageFragment(), "매물 사진 선택")
        binding.viewPager.adapter = adapter
    }

    private fun hideKeyboard() {
        if (currentFocus != null) {
            val manager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            manager.hideSoftInputFromWindow(currentFocus!!.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }

    private fun toNextPage() {
        hideKeyboard()
        val current = binding.viewPager.currentItem
        val listener = adapter.getItem(current) as OnSaveDataListener
        if (listener.checkData()) {
            listener.saveData()
            binding.viewPager.currentItem = current + 1
        } else {
            Toast.makeText(applicationContext, "error", Toast.LENGTH_SHORT).show()
        }
    }

    private fun askSave() {
        hideKeyboard()
        val current = binding.viewPager.currentItem
        val listener = adapter.getItem(current) as OnSaveDataListener
        if (listener.checkData()) {
            listener.saveData()

            val builder = AlertDialog.Builder(this)
            builder.setMessage("저장하시겠습니까?")
            builder.setIcon(R.drawable.ic_round_help_24)
            builder.setPositiveButton("확인") { _: DialogInterface, _: Int ->
                sharedViewModel.house.thumbnail = sharedViewModel.thumbnailTitle

                count = 0

                if (mode == "add") {
                    viewModel.add(sharedViewModel.house, sharedViewModel.imageUris, sharedViewModel.images)

                } else if (mode == "modify") {
                    viewModel.modify(sharedViewModel.house, sharedViewModel.imageUris, sharedViewModel.images)
                }
            }
            builder.setNegativeButton("취소") { dialog: DialogInterface, _: Int -> dialog.dismiss() }
            val alertDialog = builder.create()
            alertDialog.show()

        } else {
            Toast.makeText(applicationContext, "error", Toast.LENGTH_SHORT).show()
        }
    }
}